### PR TITLE
Fix: Cannot set property of  which has only a getter

### PR DESCRIPTION
**Pull request by dalcib**
When a method of a Class makes changes in a cloned copy of a getter property wich type is an array, the property is included in the drafts, causing the TypeError: Cannot set property 'propName' of #<Class> which has only a getter. For example:
```javascript
class State {
  [immerable] = true
  word1 = 'bar'
  word2 = 'foo'
  get chars() {
    return this.word1.concat(this.word2).split('')
  }
  setWord(newWord) {
    let mix = [...this.chars].slice(0,3)
    // The next line changes a cloned copy of this.chars
    mix[2] = 'z'
    this.word1 = mix.slice(0, 3).join('')
    this.word2 = newWord
  }
}

const state = new State()

test('should allow getter array', () => {
  expect(state.word1).toEqual('bar')
  expect(state.chars).toEqual(['b', 'a', 'r', 'f', 'o', 'o'])
  const newState = produce(state, d => d.setWord())  //TypeError: Cannot set property chars of #<State> which has only a getter
  expect(newState.word2).toEqual('baz')
})
```

This PR fixes it, checking with `ownKeys` if the property exists in the state.base, before creating the proxy. If the property doesn't exist, it returns the value.

https://codesandbox.io/s/great-colden-1py9k

### CODE REVIEW

**Comment by dalcib** [comment]
This PR also solves #439

**Comment by mweststrate** [comment]
Thanks for the suggested fix! Could you also include a unit test that makes sure we don't regress on this and the code is covered? Thanks!

**Comment by dalcib** [comment]
The tests were included @mweststrate. Thanks.

File: __tests__/base.js [hunk 1]
```diff
@@ -1147,6 +1147,47 @@ function runBaseTest(name, useProxies, autoFreeze, useListener) {
...
+		const state = new State()
+
+		it("should work without creating a proxy for a getter property", () => {
+			expect(state.chars).toEqual(['b', 'a', 'r'])
+			const newState1 = produce(state, d => d.setWord2())
+			expect(newState1.word2).toEqual('baz')
+
+			expect(state.bar).toEqual({ baz: 1 })
+			const newState2 = state.syncFoo()
+			expect(newState2.foo).toEqual(1)
```
Comment (mweststrate) [inline __tests__/base.js]: I'd check here as well if `newState2.bar` output is correct (in a case were it differs from newState1), to verify that `this.` works everywhere as expected in those getters

File: src/proxy.js [hunk 2]
```diff
@@ -130,6 +131,10 @@ function get(state, prop) {
 		// Store drafts on the copy (when one exists).
 		drafts = state.copy
 	}
+	
+	if (!ownKeys(state.base).includes(prop)) {
```
Comment (mweststrate) [inline src/proxy.js]: Note that you are checking for enumerability here, not for getter-ness. Those two are in principle unrelated, excepti that getters are by default non-enumerable. 

To find getters and setters you need check `Object.getOwnPropertyDescriptor(object, propname)`, and see if that has a `get` or `set` method.

File: src/proxy.js [hunk 3]
```diff
@@ -131,6 +131,10 @@ function get(state, prop) {
 		drafts = state.copy
 	}
 
+	if (!Object.keys(state.base).includes(prop)) {
```
Comment (dalcib) [inline src/proxy.js]: `Object.getOwnPropertyDescriptor(state.copy || state.base, prop)` always returns `undefined` for getters and methods. So, to fix the problem, it's enough to check where the property exist or not in the state.

File: __tests__/base.js (hunk 1 above)
Comment (dalcib) [reply __tests__/base.js]: Done

**Comment by dalcib** [comment]
Hi @mweststrate. The use of getOwnPropertyDescriptor didn't work, because it always returns undefined for getters. I am proposing to use Object.keys() instead.

**Comment by mweststrate** [comment]
Ah, that is because they probably live on the prototype, not the instance. I think we should hold this a little, and think this further through or maybe even forbid it. This is a rabbit hole that might go quite deep, see #317 (and #392) for some background. 

I am not sure there is an universally applicable solution to this.

**Comment by dalcib** [comment]
@mweststrate, now that v. 5.0.0 is ready, can we think about this issue?
This problem only occurs with `setUseProxies(true)`.  With es5 (without proxies) the getters work nicely. You can check it out at https://codesandbox.io/s/useimmer-tic-tac-ci6qq.
There is no performance or bundle size cost in the fix.
It allows the same behavior, with and without proxies.
@luncheon

File: src/proxy.js [hunk 4]
```diff
@@ -107,6 +107,14 @@ const objectTraps = {
...
 			drafts = state.copy
 		}
+		
+		if (!Object.keys(state.base).includes(prop)) {
+			return value
+		}
+
+		if (!Object.keys(state.base).includes(prop)) {
+			return value
+		}
```
Comment (Venryx) [inline src/proxy.js]: Are my eyes deceiving me, or are there two exact copies of the keys-check + return code? (typo?)

Comment (dalcib) [reply src/proxy.js]: Thanks

**Comment by dalcib** [comment]
Replaced by https://github.com/immerjs/immer/pull/558

**Pull request closed without merging**